processor/src/test/java/org/mapstruct/ap/test/nestedbeans/mixed/AutomappingAndNestedTest.java

```diff
@@ -200,7 +204,7 @@ public void shouldAutomapAndHandleSourceAndTargetPropertyNestingAndExpresion() {
 
         assertThat( target.getQuality() ).isNotNull();
         assertThat( target.getQuality().getReport() ).isNotNull();
-        assertThat( target.getQuality().getReport().getVerdict() ).isEqualTo( "PASSED" );
```

Comment:
That's the idea.. just pass the popped MappingOptions to the property mapping. Mapstruct will forge a new BeanMapping with these popped MappingOptions. 

Considered from a `BeanMapping` method perspective :

If the top of the stack is reached (popping yields null), its either a 
1. plain `a` to `b` mapping or nested source mapping `a.a1` to `b` (handled by property mapping itself).
2. name based / parameter based mapping 
3. or a nested target `a` to `b.b1` mapping (for that one the name based / parameter based mapping needs to stop at 'b' level', but resumes when you map `a` to `b1` which is, at that recurrence level exactly case 1. again.

Hope.. I make some sense..

 

If there's something defined on a certain level, the BeanMapping will consider it. For the remainder: name based property mapping + forging auto mappings when its a bean.